### PR TITLE
bastion configvalidator

### PR DESCRIPTION
**Pull request by tedteng**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement
/platform gcp

**What this PR does / why we need it**:
similar to https://github.com/gardener/gardener/blob/master/docs/extensions/infrastructure.md#configvalidator-interface
some of the bastion config values fetch from InfrastructureStatus, implement the validity check, whether those values exist before creating bastion VM.  eg vpc, subnet

one example from the issue cluster https://github.com/gardener/gardener-extension-provider-alicloud/issues/499#:~:text=one%20example%20from%20the%20issue%20cluster%2C%20vpc%20not%20exist%20from%20alicloud%2C%20but%20is%20saved%20in%20InfrastructureStatus

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
implement GCP bastion config validator
```

### CODE REVIEW

**Comment by tedteng** [comment]
PR is ready for review now Thanks.  @gardener/gardener-extension-provider-gcp-maintainers

File: pkg/gcp/client/compute.go [hunk 1]
```diff
@@ -32,6 +32,10 @@ import (
 type ComputeClient interface {
 	// GetExternalAddresses returns a list of all external IP addresses mapped to the names of their users.
 	GetExternalAddresses(ctx context.Context, region string) (map[string][]string, error)
+	// GetVPC returns a name of VPC
+	GetVPC(ctx context.Context, network string) (*compute.Network, error)
+	// GetSubenet returns a subnet info.
+	GetSubenet(ctx context.Context, region, subnet string) (*compute.Subnetwork, error)
```
Comment (kon-angelo) [inline pkg/gcp/client/compute.go:38]: Typo on `GetSubenet` -> `GetSubnet`

File: pkg/controller/bastion/configvalidator.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+func (c *configValidator) validateInfrastructureStatus(ctx context.Context, computeClient gcpclient.ComputeClient, region string, infrastructureStatus *gcp.InfrastructureStatus) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	vpc, err := computeClient.GetVPC(ctx, infrastructureStatus.Networks.VPC.Name)
+	if err != nil || vpc == nil || vpc.Name == "" {
+		allErrs = append(allErrs, field.InternalError(field.NewPath("vpc"), fmt.Errorf("could not get vpc %s from gcp provider: %w", infrastructureStatus.Networks.VPC.Name, err)))
+		return allErrs
+	}
+
+	subnet, err := computeClient.GetSubenet(ctx, region, infrastructureStatus.Networks.Subnets[0].Name)
```
Comment (kon-angelo) [inline pkg/controller/bastion/configvalidator.go:120]: You shouldn't use the subnet by slice index, rather the subnet with purpose `nodes`. There isn't a guarantee that this is the first will be the worker subnet and not the internal one.

**Review by kon-angelo — CHANGES_REQUESTED** [review]

File: pkg/gcp/client/compute.go (hunk 1 above)
Comment (tedteng) [reply pkg/gcp/client/compute.go]: Thanks, commit in https://github.com/gardener/gardener-extension-provider-gcp/pull/497/commits/ea08d3c63e779f44c1644aa3e6567abe3bab1c62 @kon-angelo

File: pkg/controller/bastion/configvalidator.go (hunk 2 above)
Comment (tedteng) [reply pkg/controller/bastion/configvalidator.go]: Thanks, commit in https://github.com/gardener/gardener-extension-provider-gcp/pull/497/commits/ea08d3c63e779f44c1644aa3e6567abe3bab1c62 @kon-angelo

**Comment by tedteng** [comment]
PR ready and feel free to review it thanks. @kon-angelo

**Comment by tedteng** [comment]
PR ready and feel free to review it thanks. @kon-angelo

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**